src/qibo/models/circuit.py

```diff
@@ -240,16 +247,20 @@ def __add__(self, circuit):
         newcircuit = self.__class__(**self.init_kwargs)
         # Add gates from `self` to `newcircuit` (including measurements)
         for gate in self.queue:
-            newcircuit.add(gate)
+            if isinstance(gate, gates.M):
+                newcircuit.add(gates.M(*list(gate._target_qubits)))
+            else:
+                newcircuit.add(gate)
```

Comment:
Why is this now needed? Is it associated in any way to the initialization?

Reply:
These two lines should solve the following issue:
```python
from qibo.models import Circuit
from qibo import gates
c1 = Circuit(3)
c1.add(gates.M(0,1,2,basis=gates.Y))
c2 = Circuit(3)
c = c1+c2
print(c.draw())
# Output:
# q0: ─U─U─M─
# q1: ─U─U─M─
# q2: ─U─U─M─
```
I expect to find as output something similar to the circuit `c1`, but the `U` gate is added to the circuit `c` twice because it is both in the queue of `c1`, and in the basis of `M`, so it is added also when `M` is added to `c`. It was only a temporary solution ( I forgot to put the PR in draft).

Reply:
I saw fef6a8a, and I was having hard time understanding why you wanted to avoid appending some gates.

But actually, I see nothing wrong in the snippet above: to me `+` should do the most intuitive operation, that is concatenating all the gates in `c1` to `c2`.

The only bit I do not understand is why measurement gates are floating at the end, I would have expected:
```
Output:
q0: ─U-M─U─
q1: ─U─M─U─
q2: ─U─M─U─
```

Reply:
Ah, no, I see: it is not `Circuit(3)` generating the `U`s, but `M`. Ok, now I understand the issue (why is it drawn as `U` if above you set to `Y`?)

Reply:
This practice of "adding gates" however is harmful: `M` should act as the gate, instead of adding it.
You can draw the way you prefer, but it is really weird that you ask to add a gate, and he do add two instead.

Reply:
> why is it drawn as U if above you set to Y?

Because (if I understood correctly)`basis = gate.Y` means adding to the circuit the unitary transformation into the basis of the eigenstates of `Y`, that is defined here https://github.com/qiboteam/qibo/blob/379886f075f207d81b85399ed55e11e839c87fba/src/qibo/gates/gates.py#L235-L239

Reply:
Ok, at least that is clear now.

I guess it is added to standardize the `M` gate, such that it always measures in the Z basis, simply you rotate before such that the basis as a different meaning.
But I really believe this should be internal to the `M` gate, and even drawing it is a bit confusing...

Reply:
> This practice of "adding gates" however is harmful: `M` should act as the gate, instead of adding it. You can draw the way you prefer, but it is really weird that you ask to add a gate, and he does add two instead.

I agree, maybe we could define a function `change_basis` that return the circuit to perform the basis transformation from `Z` basis into the others, in this way we could call it to evaluate `circuit_basis` (in `initialize`), and when we want to measure in another basis we append before the measurement the inverse of `change_basis`.

Reply:
> This practice of "adding gates" however is harmful: `M` should act as the gate, instead of adding it. You can draw the way you prefer, but it is really weird that you ask to add a gate, and he does add two instead.

I agree, maybe we could define a function similar to `initialize`, that returns the circuit to append before the measurements and drop the automatic appending of the unitary gates.